app/src/main/java/org/mozilla/focus/home/pocket/Pocket.kt

```diff
@@ -0,0 +1,121 @@
+import kotlinx.coroutines.experimental.Job
+import kotlinx.coroutines.experimental.async
+import kotlinx.coroutines.experimental.delay
+import kotlinx.coroutines.experimental.launch
+import org.json.JSONException
+import org.json.JSONObject
+import java.util.concurrent.TimeUnit
+
+/**
+ * Time between updates to the cache. The video feed is updated once an hour and the network
+ * API says we should cache the data for 30 minutes: for simplicity, we hardcode instead of getting
+ * it from the server and we split the difference so updates aren't too frequent.
+ */
+private val CACHE_UPDATE_FREQUENCY_MILLIS = TimeUnit.MINUTES.toMillis(45)
+
+/**
+ * A public accessor for getting the Pocket data. This class has the following responsibilities:
+ * - Public API access point
+ * - Cache container
+ * - Cache update scheduler & updater
+ *
+ * We cache in memory only. We chose not to cache on disk because it adds complexity and the cache
+ * is likely to be severely outdated between runs of the application anyway.
+ */
+object Pocket {
+
+    @Volatile private lateinit var videosCache: Deferred<List<PocketVideo>?>
+    @Volatile private var lastUpdateMillis = -1L
+    private val nextUpdateMillis get() = lastUpdateMillis + CACHE_UPDATE_FREQUENCY_MILLIS
+
+    @get:UiThread
+    @set:UiThread
+    private var backgroundUpdates: Job? = null
+
+    @JvmStatic
+    fun init() {
+        // We set this now, rather than waiting for the background updates, to ensure the first
+        // caller gets a Deferred they can wait on, rather than null (which they can't wait on).
+        videosCache = getRecommendedVideosNoCacheAsync()
+        lastUpdateMillis = SystemClock.elapsedRealtime()
+    }
+
+    @UiThread // update backgroundUpdates.
+    fun startBackgroundUpdates() {
+        backgroundUpdates?.cancel(CancellationException("Cancelling unexpectedly active job to ensure only one is running"))
+        backgroundUpdates = startBackgroundUpdatesInner()
+    }
+
+    @UiThread // update backgroundUpdates.
+    fun stopBackgroundUpdates() {
+        backgroundUpdates?.cancel(CancellationException("Stopping background updates"))
+        backgroundUpdates = null
+    }
+
+    /**
+     * Gets the recommended Pocket videos from cache. The returned [Deferred] will always contain a
+     * value except on start-up, when the initial request may still be in process. The returned value
+     * may be null which occurs when there's been an error setting an initial value in the feed.
+     *
+     * The videos are updated asynchronously so calling this twice in succession may return
+     * different lists.
+     *
+     * This method assumes [startBackgroundUpdates] has been called before being accessed.
+     */
+    @AnyThread // videosCache is synchronized.
+    fun getRecommendedVideos() = videosCache
+    private fun getRecommendedVideosNoCacheAsync() = async { PocketEndpoint.getRecommendedVideos() }
+
+    private fun startBackgroundUpdatesInner() = launch {
```

Comment:
This seems fine, launching a coroutine and using `cancel` to stop it when we're backgrounded.

Do you need any error handling, exception catching?

Reply:
afaik, I caught all the exceptions and return null in those cases so no, I think we've handled all the exception handling. That being said, I could have been more thorough and wrote tests for these error cases to ensure we actually return null for all of them.